plugin-packs/postcss-bundler/src/postcss-import/lib/parse-styles.ts

```diff
@@ -40,42 +40,37 @@ export async function parseStyles(
 			jobs.push(resolveImportId(result, stmt, postcss, require, sourceFile, base));
 		}
 
-		await Promise.all(jobs);
+		if (jobs.length) {
+			await Promise.all(jobs);
+		}
 	}
 
-	let charset: CharsetStatement | null = null;
-	const imports: Array<Statement> = [];
-	const bundle: Array<Statement> = [];
-
-	// squash statements and their children
-	statements.forEach(stmt => {
-		if (isCharsetStatement(stmt)) {
-			charset = handleCharset(stmt, charset);
-		} else if (isImportStatement(stmt)) {
-			if (stmt.children) {
-				stmt.children.forEach((child) => {
-					if (isImportStatement(child)) {
-						imports.push(child);
-					} else if (isCharsetStatement(child)) {
-						charset = handleCharset(child, charset);
-					} else {
-						bundle.push(child);
-					}
-				});
-			} else {
-				imports.push(stmt);
+	for (let i = 0; i < stylesheet.statements.length; i++) {
+		const stmt = stylesheet.statements[i];
+
+		if (isImportStatement(stmt) && stmt.stylesheet) {
+			if (stylesheet.charset && stmt.stylesheet.charset && stylesheet.charset.params.toLowerCase() !== stmt.stylesheet.charset.params.toLowerCase()) {
+				throw stmt.stylesheet.charset.error(
+					'Incompatible @charset statements:\n' +
+					`  ${stmt.stylesheet.charset.params} specified in ${stmt.stylesheet.charset.source?.input.file}\n` +
+					`  ${stylesheet.charset.params} specified in ${stylesheet.charset.source?.input.file}`,
+				);
+			} else if (!stylesheet.charset && !!stmt.stylesheet.charset) {
+				stylesheet.charset = stmt.stylesheet.charset;
 			}
-		} else if (stmt.type === 'nodes') {
-			bundle.push(stmt);
+
+			stylesheet.statements.splice(i, 1, ...stmt.stylesheet.statements);
+			i--;
+			continue;
 		}
-	});
+	}
 
-	return charset ? [charset, ...imports.concat(bundle)] : imports.concat(bundle);
```

Comment:
This is what caused the re-ordering of nodes.

`imports` and `bundle` each contained nodes that should have been interleaved.

`postcss-import` assigns each to different "buckets" and then concats these, effectively shifting up all `@import` statements.